### PR TITLE
Test that avro and JSON encoding work as expected

### PR DESCRIPTION
**Pull request by balvig**
## Background

While I think basic idea of #85 is good, two things struck me:

- We were no longer testing that the [serialization worked as advertised ([avro](https://github.com/cookpad/streamy/blob/81c69e9f29af809c74ae2ad29d7df224ca0ee73e/test/avro_event_test.rb#L59), [json](https://github.com/cookpad/streamy/blob/81c69e9f29af809c74ae2ad29d7df224ca0ee73e/test/json_event_test.rb#L27) tests identical).
- Seems message buses shouldn't be in charge of payload serialization (they just send messages).

I also wanted to touch on how we test things, since a similar quirk was apparent in https://github.com/cookpad/streamy/pull/74 and to some extent  https://github.com/cookpad/streamy/pull/80:

I feel like "emulating" what happens in the _actual_ bus in the test bus is _not_ something we should do, ie:

https://github.com/cookpad/streamy/blob/d593269c43f0d650b1a62f1c1368ca13cc8422e4/lib/streamy/message_buses/test_message_bus.rb#L8-L12

was replicating:

https://github.com/cookpad/streamy/blob/81c69e9f29af809c74ae2ad29d7df224ca0ee73e/lib/streamy/message_buses/kafka_message_bus.rb#L16-L17

This is sort of saying "we copied the code from the bus we _actually_ use and the test is passing so it must be working", but I guess that doesn't really provide long-term confidence? 😅

## This PR

What I came to realize was that we are basically interested in two things:

- Checking whether a certain type of _event_ with certain values is published (regardless of how it gets serialized)
- Ensuring that an event is delivered as a _message_ by the _message bus_, serialized in the way we intended.

Since our original way only allowed us to check the latter (and we then found a workaround to sort of make it do the former) it led me to find a piece that could sit in between the `Event` and `MessageBus` where we could look at events before they get serialized into something the buses can send:

https://github.com/cookpad/streamy/blob/b11562090a23728bd3c4c7945edcad25902c769d/lib/streamy/dispatcher.rb#L7-L11

The test version of this `Dispatcher` then replaces the `TestMessageBus`, in that it also captures any events in an array for inspection before passing them on, but otherwise follows the same codepath as the regular one:

https://github.com/cookpad/streamy/blob/b11562090a23728bd3c4c7945edcad25902c769d/lib/streamy/test_dispatcher.rb#L7-L10

That way we can have our serialization-unaware _event_ test helpers and still have tests around the very last stage of _sending a message_.

Not sure if that makes sense? 
Either way we need to find a way to get our tests back that ensured the encoding actually works 😉 

Need to test this out in in the main project first, but wanted to submit early to get some feedback.

### CODE REVIEW

File: lib/streamy/helpers/rspec_helper.rb [hunk 1]
```diff
@@ -23,13 +21,13 @@ def expect_event(topic: kind_of(String), priority: kind_of(Symbol), key: kind_of
 
       alias expect_published_event expect_event
 
-      Streamy.message_bus = MessageBuses::TestMessageBus.new
+      Dispatcher = TestDispatcher
```
Comment (dannyd4315) [inline lib/streamy/helpers/rspec_helper.rb]:
```suggestion
      Streamy::Dispatcher = TestDispatcher
```

Comment (dannyd4315) [reply lib/streamy/helpers/rspec_helper.rb]: I merged this PR into `global-web` to make sure that the tests pass. I found out that the tests don't pass due to the above. `lib/streamy/event.rb:22` does not use the TestDispatcher in tests without this. 

Another option would be to have a dispatcher method in `Event` selecting the correct dispatcher based on the environment, however, I would think that this is not the prefered option

Comment (balvig) [reply lib/streamy/helpers/rspec_helper.rb]: Oh nice catch, and thanks for testing! 👌 

> Another option would be to have a dispatcher method in Event selecting the correct dispatcher based on the environment, however, I would think that this is not the prefered option

Yes indeed! Heh, that was actually what I had in my first attempt...

 I'm not sure though, that _might_ actually still be the preferred option...? 
When I look at this again, it looks like it's re-assigning something that should be a "constant" 😅 

For now I think your change looks good though, I will apply it to the minitest/testunit version as well 👍

Comment (dannyd4315) [reply lib/streamy/helpers/rspec_helper.rb]: Yeah, I saw that as well, probably not ideal reassigning the Dispatcher. It works at the moment, but wonder if there is a better way 🤔

Comment (dannyd4315) [reply lib/streamy/helpers/rspec_helper.rb]: Hmmm, so this now works 👍 But reassigning the constant is coming up with a warning. Will have a think about another way. We could potentially do the following in the `Event` class:

```rb
def dispatcher
  if ENV["RAILS_ENV"] == "test"
    TestDispatcher
   else
     Dispatcher
   end
end
```

However, we would need to fake the rails env in the local gem with an ENV variable in the test helper.

Comment (dannyd4315) [reply lib/streamy/helpers/rspec_helper.rb]: Actually I spoke too soon, looks like we have some failures on circle, where the events are sent within a block as a batch - https://circleci.com/gh/cookpad/global-web/53142#tests/containers/1

**Comment by lewispb** [comment]
Not sure if this is still up for review or are we planning on updating to avoid reassigning the constant?

**Comment by dannyd4315** [comment]
Added a few comments, wasn't sure if you have seen them yet @balvig?

**Comment by balvig** [comment]
Hello sirs! 

Guilty as charged, been a bit swamped these past weeks 🙇 

We should definitely find a way to avoid the constant overriding 👍 

I will find my way back to it sometimes this or next week when the China dust settles :) 

(If it's blocking anything, feel free to pick up my fumble and run with it 🙇)

**Comment by balvig** [comment]
@dannyd4315 @lewispb https://github.com/cookpad/streamy/pull/87/commits/63ad9c31310a4211a1c601e91e0fa9dfb6388101 alongside #88 _mostly_ takes care of business, but as @dannyd4315 [pointed out](https://github.com/cookpad/streamy/pull/87#discussion_r284741273) there is still an issue with the "batch delivery" behavior that relies on a `TestMessageBus` pretending to be Kafka.

As we [discussed](https://github.com/cookpad/streamy/pull/74#issuecomment-484336940), I never felt completely comfortable with:

- Having the method on `Event` 
- Temporary overriding of priority
- Other message buses (ie `TestMessageBus`) now having to know about Kafka

...and I think some of that friction is showing here 🤔 

In the end for me I think it comes back to the question whether we _need_ this feature when we also have a `:batched` priority which in theory does something similar (although https://github.com/cookpad/streamy/issues/77...).

@dannyd4315 @lewispb @ChrisBr Is there any chance you could help sort this out? I know @errm meant to look into #77 but has probably been swamped with China and other stuff recently!

I _think_ ideal situation is moving away from having two different ways of sending batched events (with one having the ability to override the other), and making sure `priority :batched` works as advertised, ie reverting from

```rb
# "Manual" batch delivery
Avro::SeenFeedItemEvent.deliver do |event|
  seen_feed_items.each do |seen_feed_item|
    event.publish(user_id: user.id, seen_feed_item: seen_feed_item)
   end
end
```

```rb
# This type of event is meant to be sent in batches
module Avro
  class SeenFeedItemEvent < BaseEvents::ApplicationEvent
    priority :batched
  end
end

seen_feed_items.each do |seen_feed_item|
  event.publish(user_id: user.id, seen_feed_item: seen_feed_item)
end
```

**Comment by dannyd4315** [comment]
I have reviewed and attempted to fix the batched implementation in this PR - https://github.com/cookpad/streamy/pull/93

I have provided a setting in config now, whereby you can set a `producer_batched_message_limit`, this ensures that the messages are sent in batches of `1000` as default. This can be set in the config. 

 The `producer_batched_message_limit` should always be set below `max_buffer_size` otherwise buffer overflow errors will occur

**Comment by dannyd4315** [comment]
I have now merged https://github.com/cookpad/streamy/pull/93 into master and rebased this branch. There were a few conflicts that I fixed up, all looks good now. Going to test this branch against `global-web` 👍

**Comment by dannyd4315** [comment]
So having merged the fix for the `batched` priority and retested, I have rebased, and found an issue with the above PR. Previously all our events in global-web were using an async producer, meaning that our tests passed because we required no acknowledgment from kafka that the message was sent. However we are using `essential` priority now, so the events need an ack from Kafka that the message was received. https://github.com/cookpad/streamy/blob/cdab7a2d53228abf059f04979d70fccf89fff3a8/lib/streamy/test_dispatcher.rb

The current test dispatcher calls `super` within the dispatch method which means that it will try and produce the message to kafka and wait for an acknowledgment. I am investigating what the best way around this is. I am contemplating stubbing any requests to kafka:9092.

**Comment by dannyd4315** [comment]
Once this PR and the documentation PR are merged, we can update the streamy version and update `global-web` with the new version 👍

**Comment by dannyd4315** [comment]
♻️

File: test/event_test.rb [hunk 2]
```diff
@@ -56,15 +56,5 @@ def test_overidden_priority
...
-
-    def test_wrapping_message_bus_errors
-      Streamy.message_bus.stubs(:deliver).raises("ConnectionError")
-
-      error = assert_raises(PublicationFailedError) do
-        ValidEvent.publish
-      end
-
-      assert_match(/ConnectionError/, error.message)
-    end
```
Comment (balvig) [inline test/event_test.rb]: This behavior is tested here:

https://github.com/cookpad/streamy/blob/302776474d12c988f60790d5eecf3a89f420f1d2/test/avro_event_test.rb#L67-L79

**Comment by dannyd4315** [comment]
@balvig Your last commit is exactly what I was thinking, just testing on `global-web` now 👍

**Comment by dannyd4315** [comment]
Tested with `global-web` 👍

**Pull request merged**